keystorage/yubihsm2/src/main/java/tech/pegasys/signers/yubihsm2/YubiHsm2.java

```diff
@@ -0,0 +1,120 @@
+ */
+package tech.pegasys.signers.yubihsm2;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeoutException;
+
+import org.apache.commons.io.IOUtils;
+import org.zeroturnaround.exec.ProcessExecutor;
+import org.zeroturnaround.exec.ProcessResult;
+
+/**
+ * Access secrets/opaque data from Yubi HSM 2. https://developers.yubico.com/YubiHSM2/. It is
+ * assumed that yubi hsm sdk is installed on the system which contains yubihsm-shell. This class
+ * uses yubihsm-shell, which is a thin wrapper over libyubihsm.
+ */
+public class YubiHsm2 {
+  private final List<String> yubiHsmShellPathArgs;
+  private final String connectorUrl;
+  private final short authKey;
+  private final String password;
+  private final Optional<String> caCert;
+  private final Optional<String> proxy;
+  private final Optional<Map<String, String>> additionalEnvVars;
+
+  /**
+   * Parameters that will be passed to yubihsm-shell process to fetch opaque data
+   *
+   * @param yubiHsmShellPathArgs Path to yubihsm-shell binary and any additional arguments to pass.
+   * @param additionalEnvVars Optional environment variables map that can be passed to yubihsm-shell
+   * @param connectorUrl YubiHSM connector URL. http://127.0.01:12345 or yhusb://serial=123456
+   * @param authKeyObjId The auth key object id
+   * @param password Password
+   * @param caCert Optional CA Cert if connector is running in https mode
+   * @param proxy Optional proxy url if a proxy should be used to access connector url
+   */
+  public YubiHsm2(
+      final List<String> yubiHsmShellPathArgs,
+      final Optional<Map<String, String>> additionalEnvVars,
+      final String connectorUrl,
+      final short authKeyObjId,
+      final String password,
+      Optional<String> caCert,
+      Optional<String> proxy) {
+    this.yubiHsmShellPathArgs = yubiHsmShellPathArgs;
+    this.additionalEnvVars = additionalEnvVars;
+    this.connectorUrl = connectorUrl;
+    this.authKey = authKeyObjId;
+    this.password = password;
+    this.caCert = caCert;
+    this.proxy = proxy;
+  }
+
+  /**
+   * Fetch opaque data from yubi hsm 2
+   *
+   * @param opaqueObjId The object id of opque
+   * @return The data stored against opaque object id (in hex format)
+   * @throws YubiHsmException In case of errors in invoking process or if command returns error.
+   */
+  public String fetchKey(final short opaqueObjId) throws YubiHsmException {
+    try {
+      return executeProcess(
+          getOpaqueDataArgs(opaqueObjId),
+          additionalEnvVars.orElse(Collections.emptyMap()),
+          password);
+    } catch (final IOException | TimeoutException | InterruptedException e) {
+      throw new YubiHsmException("Error in invoking yubihsm-shell process: " + e.getMessage());
+    }
+  }
+
+  static String executeProcess(
```

Comment:
is this used somewhere else? If not - there's no need to pass in the password and EnvVars, as they are part of this object (given that the cmdline args creation is not static, this probably doesn't need to be either).

Reply:
Used in test ... I can potentially move it to a util class.

Reply:
moved into a util class. Also added a comment that it is visible for testing. As guava is not on module's dependency, can't add the @VisibleForTest annotation.